vnext/Microsoft.ReactNative/JsiApi.cpp

```diff
@@ -58,137 +43,212 @@ struct JsiBufferWrapper : facebook::jsi::Buffer {
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::WeakObject &AsWeakObject(facebook::jsi::Runtime::PointerValue **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::WeakObject *>(ptr);
-}
+  static JsiPropertyIdRef MakeJsiPropertyNameIdData(facebook::jsi::PropNameID &&propertyNameId) {
+    auto &&accessor = AsPointerAccessor(std::move(propertyNameId));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::Array const &AsArray(facebook::jsi::Runtime::PointerValue const **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::Array const *>(ptr);
-}
+  static JsiObjectRef MakeJsiArrayData(facebook::jsi::Array &&arr) {
+    auto &&accessor = AsPointerAccessor(std::move(arr));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::ArrayBuffer const &AsArrayBuffer(facebook::jsi::Runtime::PointerValue const **ptr) noexcept {
-  return *reinterpret_cast<facebook::jsi::ArrayBuffer const *>(ptr);
-}
+  static JsiWeakObjectRef MakeJsiWeakObjectData(facebook::jsi::WeakObject &&weakObj) {
+    auto &&accessor = AsPointerAccessor(std::move(weakObj));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
 
-static facebook::jsi::Value const *AsValue(JsiValueRef const &data) noexcept {
-  return reinterpret_cast<facebook::jsi::Value const *>(&data);
-}
+  static JsiObjectRef MakeJsiFunctionData(facebook::jsi::Function &&func) {
+    auto &&accessor = AsPointerAccessor(std::move(func));
+    return {reinterpret_cast<uint64_t>(std::exchange(accessor.ptr_, nullptr))};
+  }
+};
 
-static JsiValueRef ToJsiValueData(facebook::jsi::Value const &value) noexcept {
-  return {static_cast<JsiValueKind>(static_cast<int32_t>(value.kind_)),
-          *reinterpret_cast<uint64_t const *>(&value.data_.number)};
-}
+struct ValueAccessor {
+  static JsiValueRef ToJsiValueData(facebook::jsi::Value const &value) noexcept {
+    ValueAccessor const &accessor = reinterpret_cast<ValueAccessor const &>(value);
+    return {static_cast<JsiValueKind>(static_cast<int32_t>(accessor.m_kind)),
+            *reinterpret_cast<uint64_t const *>(&accessor.m_data)};
+  }
 
-static JsiPropertyIdRef ToJsiPropertyNameIdData(facebook::jsi::PropNameID const &propNameId) noexcept {
-  return {reinterpret_cast<uint64_t>(propNameId.ptr_)};
-}
+  static JsiValueRef MakeJsiValueData(facebook::jsi::Value &&value) {
+    ValueAccessor &&accessor = reinterpret_cast<ValueAccessor &&>(value);
+    return {static_cast<JsiValueKind>(static_cast<int32_t>(std::exchange(accessor.m_kind, UndefinedKind))),
+            *reinterpret_cast<uint64_t *>(&accessor.m_data)};
+  }
 
-static JsiSymbolRef MakeJsiSymbolData(facebook::jsi::Runtime::PointerValue *pointerValue) {
-  return {reinterpret_cast<uint64_t>(pointerValue)};
-}
+ private:
+  enum ValueKind { UndefinedKind } m_kind;
+  double m_data;
```

Comment:
should this just be a void* ?

Reply:
No, must not use double to make sure that it works for x86. The data_ member in facebook::jsi::Value is an enum that has double, pointer, etc. I put here `double` because it has the most strict requirements for the value size and alignment. In x86 void* would be 4 bytes instead of expected 8 bytes.